MSBuild inconsistent file path to the same Task assembly load casing cast error

### Steps to reproduce

Failed to build https://github.com/dotnet-maestro-bot/cli/tree/merge/release/3.1.1xx-to-release/3.1.2xx due to error:

```
C:\work\cli\src\Microsoft.DotNet.MSBuildSdkResolver\Microsoft.DotNet.MSBuildSdkResolver.csproj]
C:\work\cli\.packages\microsoft.build.tasks.git\1.0.0-beta2-19367-01\buildCrossTargeting\Microsoft.Build.Tasks.Git.targets(24,5): error MSB4018: System.InvalidCastException: [A]System.Runtime.CompilerServices.StrongBox`1[Microsoft.Build.Tasks.Git.GitRepository] cannot be cast to [B]System.Runtime.CompilerServices.StrongBox`1[Microsoft.Build.Tasks.Git.GitRepository]. Type A originates from 'System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e' in the context 'Default' at location 'C:\work\cli\.dotnet\shared\Microsoft.NETCore.App\3.1.2\System.Private.CoreLib.dll'. Type B originates from 'System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e' in the context 'Default' at location 'C:\work\cli\.dotnet\shared\Microsoft.NETCore.App\3.1.2\System.Private.CoreLib.dll'. [C:\work\cli\src\Microsoft.DotNet.MSBuildSdkResolver\Microsoft.DotNet.MSBuildSdkResolver.csproj]
C:\work\cli\.packages\microsoft.build.tasks.git\1.0.0-beta2-19367-01\buildCrossTargeting\Microsoft.Build.Tasks.Git.targets(24,5): error MSB4018:    at Microsoft.Build.Tasks.Git.RepositoryTask.TryGetCachedRepositoryInstance(String cacheKey, Boolean requireCached, GitRepository& repository) in /_/src/Microsoft.Build.Tasks.Git/RepositoryTask.cs:line 135 [C:\work\cli\src\Microsoft.DotNet.MSBuildSdkResolver\Microsoft.DotNet.MSBuildSdkResolver.csproj]
```

The same assembly with different path is loaded
![image](https://user-images.githubusercontent.com/6993335/73213652-68f8d000-4105-11ea-93cd-627a1ed6c837.png)


It is caused by path in https://github.com/microsoft/msbuild/blob/86d9494e446de002ea6a4f08d6fbd111b1041059/src/Shared/CoreCLRAssemblyLoader.cs#L32 is not normalized.

--- a/src/Build.UnitTests/BackEnd/AssemblyLoadContextTestTasks.cs
+++ b/src/Build.UnitTests/BackEnd/AssemblyLoadContextTestTasks.cs
@@ -1,0 +1,36 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace AssemblyLoadContextTest
+{
+    public class RegisterObject : Task
+    {
+        internal const string CacheKey = "RegressionForMSBuild#5080";
+
+        public override bool Execute()
+        {
+            BuildEngine4.RegisterTaskObject(
+                  CacheKey,
+                  new RegisterObject(),
+                  RegisteredTaskObjectLifetime.Build,
+                  allowEarlyCollection: false);
+
+            return true;
+        }
+    }
+
+    public class RetrieveObject : Task
+    {
+        public override bool Execute()
+        {
+            var entry = (RegisterObject)BuildEngine4.GetRegisteredTaskObject(RegisterObject.CacheKey, RegisteredTaskObjectLifetime.Build);
+
+            return true;
+        }
+    }
+}

--- a/src/Build.UnitTests/BackEnd/TaskBuilder_Tests.cs
+++ b/src/Build.UnitTests/BackEnd/TaskBuilder_Tests.cs
@@ -517,6 +517,38 @@ namespace Microsoft.Build.UnitTests.BackEnd
             logger.AssertLogContains("[foo: ]");
         }
 
+        /// <summary>
+        /// Regression test for https://github.com/microsoft/msbuild/issues/5080
+        /// </summary>
+        [Fact]
+        public void SameAssemblyFromDifferentRelativePathsSharesAssemblyLoadContext()
+        {
+            string realTaskPath = Assembly.GetExecutingAssembly().Location;
+
+            string fileName = Path.GetFileName(realTaskPath);
+            string directoryName = Path.GetDirectoryName(realTaskPath);
+
+            using var env = TestEnvironment.Create();
+
+            string customTaskFolder = Path.Combine(directoryName, "buildCrossTargeting");
+            env.CreateFolder(customTaskFolder);
+
+            string projectContents = @"<Project ToolsVersion=`msbuilddefaulttoolsversion` xmlns=`msbuildnamespace`>
+  <UsingTask TaskName=`RegisterObject` AssemblyFile=`" + Path.Combine(customTaskFolder, "..", fileName) + @"` />
+  <UsingTask TaskName=`RetrieveObject` AssemblyFile=`" + realTaskPath + @"` />
+
+  <Target Name=`Build`>
+    <RegisterObject />
+    <RetrieveObject />
+  </Target>
+</Project>";
+
+            MockLogger logger = ObjectModelHelpers.BuildProjectExpectSuccess(projectContents, _testOutput);
+
+            logger.AssertLogDoesntContain("MSB4018");
+        }
+
+
 #if FEATURE_CODETASKFACTORY
         /// <summary>
         /// If an item being output from a task has null metadata, we shouldn't crash. 

--- a/src/Shared/CoreCLRAssemblyLoader.cs
+++ b/src/Shared/CoreCLRAssemblyLoader.cs
@@ -26,6 +26,11 @@ namespace Microsoft.Build.Shared
 
             Debug.Assert(Path.IsPathRooted(fullPath));
 
+            // Normalize because the same assembly might get loaded via
+            // multiple paths (for instance, the `build` and `buildCrossTargeting`
+            // folders in a NuGet package).
+            fullPath = FileUtilities.NormalizePath(fullPath);
+
             lock (_guard)
             {
                 Assembly assembly;

